synapse/handlers/space_summary.py

```diff
@@ -111,28 +116,75 @@ async def get_space_summary(
 
             if is_in_room:
-                rooms, events = await self._summarize_local_room(
-                    requester, room_id, suggested_only, max_children
+                room, events = await self._summarize_local_room(
+                    requester, None, room_id, suggested_only, max_children
                 )
+
+                logger.debug(
+                    "Query of local room %s returned events %s",
+                    room_id,
+                    ["%s->%s" % (ev["room_id"], ev["state_key"]) for ev in events],
+                )
+
+                if room:
+                    rooms_result.append(room)
             else:
-                rooms, events = await self._summarize_remote_room(
+                fed_rooms, fed_events = await self._summarize_remote_room(
                     queue_entry,
                     suggested_only,
                     max_children,
                     exclude_rooms=processed_rooms,
                 )
 
-            logger.debug(
-                "Query of %s returned rooms %s, events %s",
-                queue_entry.room_id,
-                [room.get("room_id") for room in rooms],
-                ["%s->%s" % (ev["room_id"], ev["state_key"]) for ev in events],
-            )
-
-            rooms_result.extend(rooms)
-
-            # any rooms returned don't need visiting again
-            processed_rooms.update(cast(str, room.get("room_id")) for room in rooms)
+                # The results over federation might include rooms that the we,
+                # as the requesting server, are allowed to see, but the requesting
+                # user is not permitted see.
+                #
+                # A room may be visible if it is world readable, or if it is accessible
+                # to the specific user.
+                room_ids = set()
+                events = []
+                for room in fed_rooms:
+                    fed_room_id = room.get("room_id")
+                    if not fed_room_id:
+                        continue
+
+                    # Check whether the room is world readable from the response.
+                    include_room = room.get("world_readable") is True
+
+                    # If the room is not world-readable, get the allowed spaces from
+                    # the response.
+                    allowed_spaces = room.get("allowed_spaces")
+                    if not include_room and allowed_spaces:
+                        include_room = await self._event_auth_handler.is_user_in_rooms(
+                            allowed_spaces, requester
+                        )
+
+                    # Finally, check ourselves if we can access the room (this will
+                    # fail if we don't have any state for this room).
+                    if not include_room:
+                        include_room = await self._is_room_accessible(
+                            fed_room_id, requester, None
+                        )
```

Comment:
If we're not in the room this should always fail, no?

Reply:
Is this trying to check for invites? I don't think that will work.

Reply:
We call the federation API on a root room and it returns the root room + children rooms, so we *might* be in the room. (We only know we weren't in the "root" room, not the children room.) I think the comment suggested above is also incorrect for this reason, but I'm going to add one which explains this.

Reply:
I updated the comment above this to hopefully clarify what's going on.